net_connection/src/protocol.rs

```diff
@@ -0,0 +1,276 @@
+                name: b"ping".to_vec(),
+                data: rmp_serde::to_vec_named(p).unwrap(),
+            },
+            Protocol::Pong(p) => NamedBinaryData {
+                name: b"pong".to_vec(),
+                data: rmp_serde::to_vec_named(p).unwrap(),
+            },
+            Protocol::P2pReady => NamedBinaryData {
+                name: b"p2pReady".to_vec(),
+                data: Vec::new(),
+            },
+        }
+    }
+}
+
+impl From<Protocol> for NamedBinaryData {
+    fn from(p: Protocol) -> Self {
+        (&p).into()
+    }
+}
+
+/// provide utility for Protocol deserialization
+impl<'a> From<&'a NamedBinaryData> for Protocol {
+    fn from(nb: &'a NamedBinaryData) -> Self {
+        match nb.name.as_slice() {
+            b"namedBinary" => {
+                let sub: NamedBinaryData = rmp_serde::from_slice(&nb.data).unwrap();
+                Protocol::NamedBinary(sub)
+            }
+            b"json" => Protocol::Json(String::from_utf8_lossy(&nb.data).to_string().into()),
+            b"ping" => {
+                let sub: PingData = rmp_serde::from_slice(&nb.data).unwrap();
+                Protocol::Ping(sub)
+            }
+            b"pong" => {
+                let sub: PongData = rmp_serde::from_slice(&nb.data).unwrap();
+                Protocol::Pong(sub)
+            }
+            b"p2pReady" => Protocol::P2pReady,
+            _ => panic!("bad Protocol type: {}", String::from_utf8_lossy(&nb.name)),
+        }
+    }
+}
+
+impl From<NamedBinaryData> for Protocol {
+    fn from(nb: NamedBinaryData) -> Self {
+        (&nb).into()
+    }
+}
+
+impl<'a> From<&'a str> for Protocol {
+    fn from(s: &'a str) -> Self {
+        Protocol::Json(s.to_string().into())
+    }
+}
+
+impl From<String> for Protocol {
+    fn from(s: String) -> Self {
+        s.as_str().into()
+    }
+}
+
+impl From<Protocol> for String {
+    fn from(p: Protocol) -> String {
+        p.as_json_string()
+    }
+}
+
+/// local macro for creating is_* and as_* functions on Protocol
+/// (DRY some boilerplate)
+macro_rules! simple_access {
+    ($($is:ident $as:ident $d:ident $t:ty)*) => {
+        $(
+            pub fn $is(&self) -> bool {
+                if let Protocol::$d(_) = self {
+                    true
+                } else {
+                    false
+                }
+            }
+
+            pub fn $as<'a>(&'a self) -> &'a $t {
+                if let Protocol::$d(data) = self {
+                    &data
+                } else {
+                    panic!(concat!(stringify!($as), " called with bad type"));
+                }
+            }
+        )*
+    }
+}
+
+impl Protocol {
+    simple_access! {
+        is_named_binary as_named_binary NamedBinary NamedBinaryData
+        is_json as_json Json JsonString
+        is_ping as_ping Ping PingData
+        is_pong as_pong Pong PongData
+    }
+
+    /// get a json string straight out of the Protocol enum
+    pub fn as_json_string(&self) -> String {
```

Comment:
should return `JsonString` no?

there are good reasons not to put known-json data into a raw `String` primitive, documented https://github.com/holochain/holochain-rust/pull/490 and https://hackmd.io/EmRHEEpuRRihr_K7MThVZg#centralize-serialization-logic

if there is a use-case for having raw `String` values containing json floating around, i'd like to know so i can maybe extend what i've done to cover it?

Reply:
:+1: updated to JsonString